File: TheSky/lua/op/GuideView/selectRoleLayer.lua
local _layer
local _current
local _heroIds = {}
local _roleSprites = {}
local _heroDesps = {}
local _heroNames = {}
-- renzhan
local btnPositionXs = {}
local btnPositionYs = {}

local KTAG_ATTR = 987

local KTAG_BLACKLAYER = 997

local buttonLayer
-- 

local HERO_ATTACK_FRAMECOUNT = {
    hero_000338 = 33,
    hero_000304 = 30,
    hero_000303 = 32,
}

local HERO_STAND_FRAMECOUNT = {
    hero_000338 = 4,
    hero_000304 = 4,
    hero_000303 = 4,
}

-- ·名字不要重复
SelectRoleOwner = SelectRoleOwner or {}
ccb["SelectRoleOwner"] = SelectRoleOwner

local function _getActionAnimation(resId,action,frameCount)
    local animFrames = CCArray:create()
    for j = 1, frameCount do
        local frameName = string.format("sel_%s_%s_%04d.png", resId, action, j)
        
        local frame = CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName(frameName)
        animFrames:addObject(frame)
    end
    local animation = CCAnimation:createWithSpriteFrames(animFrames, 0.1)
    return animation
end


local function _selectRole(tag)
    if _current == tag then
        return
    end

    playMusic("audio/sound"..(153-tag)..".mp3",false)

    -- renzhan
    for i=1,3 do
        SelectRoleOwner["roleBtn"..i]:getChildByTag(KTAG_BLACKLAYER):setVisible(false)
        -- 禁用按钮
        SelectRoleOwner["roleBtn"..i]:setEnabled(false)
    end

    -- 缩小
    local item = SelectRoleOwner["roleBtn".._current]
    local despText = _heroDesps[_current]
    local nameLabel = _heroNames[_current]
    if item then
        item:setZOrder(1)
        item:runAction(CCScaleTo:create(0.3, retina))
        item:setNormalSpriteFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName(string.format("role_btn_%d_0.png", _current)))

        local Attr = item:getChildByTag(KTAG_ATTR)
        Attr:setVisible(false)
        despText:setVisible(true)
    end
    if despText then
        despText:runAction(CCMoveBy:create(0.3, ccp(0, 50)))
    end
    if nameLabel then
        nameLabel:runAction(CCMoveBy:create(0.3, ccp(0, -50)))
    end

    local roleSprite = _roleSprites[_current]
    if roleSprite then
        roleSprite:stopAllActions()
        local animation = _getActionAnimation(_heroIds[_current], "stand", HERO_STAND_FRAMECOUNT[_heroIds[_current]])
        roleSprite:runAction(CCRepeatForever:create(CCAnimate:create(animation)))
    end

    --  放大
    _current = tag
    item = SelectRoleOwner["roleBtn".._current]
    despText =  _heroDesps[_current]
    nameLabel = _heroNames[_current]
    if item then
        item:setZOrder(3)
        -- item:setScale(1.5 * retina)
        item:runAction(CCScaleTo:create(0.3, 1.32 * retina))
        item:setNormalSpriteFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName(string.format("role_btn_%d_1.png", _current)))

        local Attr = item:getChildByTag(KTAG_ATTR)
        Attr:setVisible(true)
        despText:setVisible(false)

    end
    if despText then
        despText:runAction(CCMoveBy:create(0.3, ccp(0, -50)))
    end
    if nameLabel then
        nameLabel:runAction(CCMoveBy:create(0.3, ccp(0, 50)))
    end

    roleSprite = _roleSprites[_current]
    if roleSprite then
        roleSprite:stopAllActions()
        local array = CCArray:create()
        local animation1 = _getActionAnimation(_heroIds[_current], "stand", HERO_STAND_FRAMECOUNT[_heroIds[_current]])
        array:addObject(CCAnimate:create(animation1))
        local animation2 = _getActionAnimation(_heroIds[_current], "stand", HERO_STAND_FRAMECOUNT[_heroIds[_current]])
        array:addObject(CCAnimate:create(animation2))  
        local animation3 = _getActionAnimation(_heroIds[_current], "attack", HERO_ATTACK_FRAMECOUNT[_heroIds[_current]])
        array:addObject(CCAnimate:create(animation3))
        roleSprite:runAction(CCRepeatForever:create(CCSequence:create(array)))
    end

    item:runAction(CCMoveTo:create(0.3,ccp(btnPositionXs[2] , btnPositionYs[2])))

    local next1 = (_current + 1) <= 3 and (_current + 1) or _current + 1 - 3
    local next2 = (_current + 2) <= 3 and (_current + 2) or _current + 2 - 3
    SelectRoleOwner["roleBtn"..next1]:runAction(CCMoveTo:create(0.3, ccp(btnPositionXs[3], btnPositionYs[3])))
    SelectRoleOwner["roleBtn"..next2]:runAction(CCMoveTo:create(0.3, ccp(btnPositionXs[1], btnPositionYs[1])))


    local function blackLayerShow( )
        for i=1,3 do
            if i ~= _current then

                SelectRoleOwner["roleBtn"..i]:getChildByTag(KTAG_BLACKLAYER):setVisible(true)
                -- 解除禁用
                SelectRoleOwner["roleBtn"..i]:setEnabled(true)
                SelectRoleOwner["roleBtn"..i]:setEnabled(true)
            end
        end
    end 

    local blackArray = CCArray:create()
    blackArray:addObject(CCDelayTime:create(0.3))
    blackArray:addObject(CCCallFunc:create(blackLayerShow))
    _layer:runAction(CCSequence:create(blackArray))

end

local function roleClick(tag)
    _selectRole(tag)
end
SelectRoleOwner["roleClick"] = roleClick

local function confirmClick()
    local heroId = _heroIds[_current]
    if heroId then
        _layer:getParent():addChild(createCreateNameLayer(heroId, -135)) 
    end
end
SelectRoleOwner["confirmClick"] = confirmClick

local function _addRoleAnimations()
    for i=1,3 do
        -- local aniRes = string.format("%s_stand_0001.png",_heroIds[i])
         local aniRes = string.format("sel_%s_stand_0001.png", _heroIds[i])
        local roleSprite = CCSprite:createWithSpriteFrameName(aniRes)
        local animation = _getActionAnimation(_heroIds[i], "stand", HERO_STAND_FRAMECOUNT[_heroIds[i]])
        local item = SelectRoleOwner["roleBtn"..i]
        item:addChild(roleSprite,4)
        roleSprite:setScaleX(-1.2)
        roleSprite:setScaleY(1.2)
        roleSprite:setPosition(ccp(item:getContentSize().width * 0.5, item:getContentSize().height * 0.5))
        roleSprite:runAction(CCRepeatForever:create(CCAnimate:create(animation)))
        if i == 1 then
            roleSprite:setFlipX(true)
        end
        table.insert(_roleSprites, roleSprite)
    end
end

local function _addRoleInfos()
    for i=1,3 do
        local item = SelectRoleOwner["roleBtn"..i]
        local despRes = _heroDesps[i]
        local despSprite = CCSprite:createWithSpriteFrameName(despRes)
        item:addChild(despSprite, 5)
        despSprite:setPosition(ccp(item:getContentSize().width * 0.5,item:getContentSize().height * 0.75))
        _heroDesps[i] = despSprite

        local nameRes = _heroNames[i]
        local nameSprite = CCSprite:createWithSpriteFrameName(nameRes)
        item:addChild(nameSprite, 5)
        nameSprite:setPosition(ccp(item:getContentSize().width * 0.5,item:getContentSize().height * 0.23))
        _heroNames[i] = nameSprite
    end
end


-- private方法如果没有上下调用关系可以写在createLayer外面
local function _init()
    -- get layer
    CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/selectRole.plist")
    CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/selectRoleRobin.plist")
    CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/selectRoleSanji.plist")
    CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/selectRoleZoro.plist")
    CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("fontPic.plist")
    _scene = CCScene:create()
    local  proxy = CCBProxy:create()
    local  node  = CCBReaderLoad("ccbResources/SelectRoleView.ccbi", proxy, true,"SelectRoleOwner")
    
    _layer = tolua.cast(node, "CCLayer")
    _scene:addChild(_layer)
    _addRoleAnimations()
    _addRoleInfos()
    SelectRoleOwner["confirmItem"]:setZOrder(10)

    for i = 1,3 do
        btnPositionXs[i] = SelectRoleOwner["roleBtn"..i]:getPositionX()
        btnPositionYs[i] = SelectRoleOwner["roleBtn"..i]:getPositionY()

        local heroConfig = herodata:getHeroConfig(_heroIds[i])
        local hp = heroConfig.attr.hp
        local atk = heroConfig.attr.atk
        local def = heroConfig.attr.def
        local mp = heroConfig.attr.mp

        local hpFrame = CCSprite:createWithSpriteFrameName("attr_bg.png")
        SelectRoleOwner["roleBtn"..i]:addChild(hpFrame, 4, KTAG_ATTR)

        local aniRes = string.format("sel_%s_stand_0001.png", _heroIds[i])

        local roleSprite = CCSprite:createWithSpriteFrameName(aniRes)
        hpFrame:setPosition(ccp(SelectRoleOwner["roleBtn"..i]:getContentSize().width * 0.5, 
            SelectRoleOwner["roleBtn"..i]:getContentSize().height * 0.71))

        CCSpriteFrameCache:sharedSpriteFrameCache():addSpriteFramesWithFile("ccbResources/publicRes.plist")
        local hpIcon = CCSprite:createWithSpriteFrameName("hp_icon.png")
        hpFrame:addChild(hpIcon)
        hpIcon:setPosition(ccp(30,hpFrame:getContentSize().height / 4 * 3))
        local hpLabel = CCLabelTTF:create(hp, "ccbResources/FZCuYuan-M03S.ttf", 25)
        hpFrame:addChild(hpLabel)
        hpLabel:setAnchorPoint(ccp(0,0.5))
        hpLabel:setPosition(ccp(48,hpFrame:getContentSize().height / 4 * 3))

        local atkIcon = CCSprite:createWithSpriteFrameName("atk_icon.png")
        hpFrame:addChild(atkIcon)
        atkIcon:setPosition(ccp(115,hpFrame:getContentSize().height / 4 * 3))
        local atkLabel = CCLabelTTF:create(atk, "ccbResources/FZCuYuan-M03S.ttf", 25)
        hpFrame:addChild(atkLabel)
        atkLabel:setAnchorPoint(ccp(0,0.5))
        atkLabel:setPosition(ccp(133,hpFrame:getContentSize().height / 4 * 3))

        local defIcon = CCSprite:createWithSpriteFrameName("def_icon.png")
        hpFrame:addChild(defIcon)
        defIcon:setPosition(ccp(30,hpFrame:getContentSize().height / 4))
        local defLabel = CCLabelTTF:create(def, "ccbResources/FZCuYuan-M03S.ttf", 25)
        hpFrame:addChild(defLabel)
        defLabel:setAnchorPoint(ccp(0,0.5))
        defLabel:setPosition(ccp(48,hpFrame:getContentSize().height / 4))

        local mpIcon = CCSprite:createWithSpriteFrameName("int_icon.png")
        hpFrame:addChild(mpIcon)
        mpIcon:setPosition(ccp(115,hpFrame:getContentSize().height / 4))
        local mpLabel = CCLabelTTF:create(mp, "ccbResources/FZCuYuan-M03S.ttf", 25)
        hpFrame:addChild(mpLabel)
        mpLabel:setAnchorPoint(ccp(0,0.5))
        mpLabel:setPosition(ccp(133,hpFrame:getContentSize().height / 4))
        hpFrame:setVisible(false)
        hpFrame:setScale(0.8)
        hpFrame:setOpacity(150)
    end

    -- 添加灰色遮罩
    for i=1,3 do
        local blackLayerBg = CCLayerColor:create(ccc4(0, 0, 0, 150),SelectRoleOwner["roleBtn"..i]:getContentSize().width,SelectRoleOwner["roleBtn"..i]:getContentSize().height)

        SelectRoleOwner["roleBtn"..i]:addChild(blackLayerBg, 6, KTAG_BLACKLAYER)
        blackLayerBg:setVisible(false)
    end
-- 
end


-- 该方法名字每个文件不要重复
function getSelectRoleLayer()
	return _layer
end

function createSelectRoleLayer()
    _heroIds[1] = "hero_000338"
    _heroIds[2] = "hero_000303"
    _heroIds[3] = "hero_000304"

    _heroDesps[1] = "robin_desp_text.png"
    _heroDesps[2] = "sanji_desp_text.png"
    _heroDesps[3] = "zoro_desp_text.png"
    _heroNames[1] = "robin_text.png"
    _heroNames[2] = "sanji_text.png"
    _heroNames[3] = "zoro_text.png"

    _init()
    _current = 0
    -- _selectRole(1)

    local function _onEnter()
    end

    local function _onExit()
        print("onExit")
        for i,v in ipairs(_roleSprites) do
            v:stopAllActions()
            v:removeFromParentAndCleanup(true)
            v = nil
        end
        _roleSprites = {}
        _heroIds = {}
        _heroDesps = {}
        _heroNames = {}
        _layer = nil
    end

    --onEnter onExit
    local function _layerEventHandler(eventType)
        if eventType == "enter" then
            if _onEnter then _onEnter() end
        elseif eventType == "exit" then
            if _onExit then _onExit() end
        end
    end
    _layer:registerScriptHandler(_layerEventHandler)
    
    return _scene
end

File: TheSky/JavaScript/LZ/scenes/GuideScene/SelectRoleLayer.js
var SelectRoleLayer = cc.Layer.extend({
	ctor:function(){
		this._super();
		this.initLayer();
	},
	onEnter:function(){
		this._super();
	},
	onExit:function(){
		this._super();
	},
	initLayer:function(){
		cc.BuilderReader.registerController("SelectRoleOwner", {
		});
		this.node = cc.BuilderReader.load(ccbi_res.SelectRoleView_ccbi, this);
		if(this.node != null) {
			this.addChild(this.node);
		};
		this.heroIds = ["hero_000401","hero_000406","hero_000419"];
		this.heroDesp = ["robin_desp_text.png","sanji_desp_text.png","zoro_desp_text.png"];
		this.heroNames = ["robin_text.png","sanji_text.png","zoro_text.png"];
		this.confirmItem.setLocalZOrder(10);
		this.addRoleInfos();
		this.refresh()
		this.current;
		for (var i = 0; i < 3; i++) {
			var blackLayerBg = cc.LayerColor.create(cc.color(0, 0, 0, 150), this["roleBtn" + (i+1)].getContentSize().width, this["roleBtn" + (i+1)].getContentSize().height);
			this["roleBtn" +  (i+1)].addChild(blackLayerBg, 16 ,10);
			blackLayerBg.visible = false;
		}
		
		
		
//		traceTable("config--", array)
	},
	addRoleInfos:function(){
		for (var i = 1; i < 4; i++) {
			var item = this["roleBtn" + i];
			var despRes = this.heroDesp[i - 1];
			var despSprite = new cc.Sprite("#" + despRes);
			item.addChild(despSprite, 5);
			despSprite.attr({
				x:item.getContentSize().width / 2,
				y:item.getContentSize().height * 0.75,
				anchorX:0.5,
				anchorY:0.5
			});
			this.heroDesp[i - 1] = despSprite;

			var nameRes = this.heroNames[i - 1];
			var nameSprite  = new cc.Sprite("#" + nameRes);
			item.addChild(nameSprite, 5);
			nameSprite.attr({
				x:item.getContentSize().width / 2,
				y:item.getContentSize().height * 0.22,
				anchorX:0.5,
				anchorY:0.5
			});
			this.heroNames[i - 1] = nameSprite;
		}
	},
	refresh:function(){

	},
	roleClick:function(sender){
		var tag = sender.getTag();
		if (tag == this.current) {
			return;
		} else {
			this.current = tag;
		}
		
		
		this.manager = ccs.armatureDataManager;
		for (var i = 0; i < 1; i++) {
			var jsonPath = "bone/guide/Tiny_CHEER/Tiny.ExportJson";
			var img = "bone/guide/Tiny_CHEER/Tiny" +  i +".png";
			var plist = "bone/guide/Tiny_CHEER/Tiny" + i + ".plist";
			this.manager.addArmatureFileInfo(img, plist, jsonPath);
		}
		this.guidePuppet = new ccs.Armature("Tiny");
		var animation = this.guidePuppet.getAnimation();
		animation.play("Cheer", -1, 0);
		this.guidePuppet.attr({
			x:cc.winSize.width / 2,
			y:cc.winSize.height / 2,
			anchorX:0.5,
			anchorY:0.5
		});
		this.guidePuppet.getAnimation().setMovementEventCallFunc(function(target, type, movementName){
			if (type > 0) {
				this.guidePuppet.removeFromParent(true);
			}
		}.bind(this), this.guidePuppet);
		this.guidePuppet.getAnimation().setFrameEventCallFunc(function(target, event, originFrameIndex, currentFrameIndex){
			if (event === "ontarget") {
			}
		}.bind(this), this.guidePuppet);
		this.guidePuppet.setScale(5);
		this["role_" +  this.current].addChild(this.guidePuppet, 100);
		
		
		
		for (var i = 0; i < 3; i++) {
			this["roleBtn" +  (i+1)].getChildByTag(10).visible = false;
			this["roleBtn" +  (i+1)].setEnabled(false);
			this.heroDesp[i].attr({
				x:this["roleBtn" +  (i+1)].getContentSize().width / 2,
				y:this["roleBtn" +  (i+1)].getContentSize().height * 0.75,
				anchorX:0.5,
				anchorY:0.5
			});
			this.heroNames[i].attr({
				x:this["roleBtn" +  (i+1)].getContentSize().width / 2,
				y:this["roleBtn" +  (i+1)].getContentSize().height * 0.22,
				anchorX:0.5,
				anchorY:0.5
			});
		}
		var item = this["roleBtn" +  this.current];
		var despText = this.heroDesp[this.current - 1];
		var nameLabel = this.heroNames[this.current - 1];
		if (item) {
			item.setLocalZOrder(1);
		}
		if (despText) {
			despText.runAction(cc.moveBy(0.3, cc.p(0, 50)));
		}
		if (nameLabel) {
			nameLabel.runAction(cc.moveBy(0.3, cc.p(0, -50)));
		}
		this.despText1.runAction(cc.sequence(
				cc.delayTime(0.3),
				cc.callFunc(function(){
					for (var i = 0; i < 3; i++) {
						if ((i + 1) != this.current) {
							this["roleBtn" +  (i+1)].getChildByTag(10).visible = true;
							this["roleBtn" +  (i+1)].setEnabled(true);
						}
					}
				}.bind(this))
		));
		
	},
	confirmClick:function(){
		var heroId = this.heroIds[this.current -1];
		if (heroId) {
			var layer = new CreateNameLayer(heroId);
			cc.director.getRunningScene().addChild(layer);
		}
	},
})